src/core/net/mdns.cpp

```diff
@@ -2614,19 +2676,35 @@ void Core::ServiceType::AppendPtrRecordTo(TxMessage &aResponse, uint16_t aServic
 //----------------------------------------------------------------------------------------------------------------------
 // Core::TxMessage
 
-Core::TxMessage::TxMessage(Instance &aInstance, Type aType)
+Core::TxMessage::TxMessage(Instance &aInstance, Type aType, uint16_t aQueryId)
     : InstanceLocator(aInstance)
 {
-    Init(aType);
+    Init(aType, aQueryId);
 }
 
-Core::TxMessage::TxMessage(Instance &aInstance, Type aType, const AddressInfo &aUnicastDest)
-    : TxMessage(aInstance, aType)
+Core::TxMessage::TxMessage(Instance &aInstance, Type aType, const AddressInfo &aUnicastDest, uint16_t aQueryId)
+    : TxMessage(aInstance, aType, aQueryId)
 {
     mUnicastDest = aUnicastDest;
 }
 
-void Core::TxMessage::Init(Type aType)
+Core::TxMessage::TxMessage(Instance          &aInstance,
+                           Type               aType,
+                           const AddressInfo &aUnicastDest,
+                           OwnedPtr<Message> &aReceivedQuery,
+                           uint16_t           aQueryId)
+    : TxMessage(aInstance, aType, aUnicastDest, aQueryId)
+{
+    uint16_t offset = sizeof(Header);
+    IgnoreError(Name(*aReceivedQuery, sizeof(Header)).AppendTo(*mMsgPtr));
+    IgnoreError(Name::ParseName(*aReceivedQuery, offset));
+    // RFC6762, section 6.7:
+    // Legacy Unicast Response must repeat the question
+    IgnoreError(mMsgPtr->AppendBytesFromMessage(*aReceivedQuery, offset, sizeof(ot::Dns::Question)));
```

Comment:
We may repeat the same name later in answer section, but I cannot think of an easy way to save the offset. Do you have ant ideas?

I would say it may be fine to not compress this name in legacy unicast responses?

Reply:
I am tempted to assume that the name found in the answer section could/will always have ```nameOffset = sizeof(Header)```. I don't know what the effort will be to adapt ```PrepareResponse()``` and ```PrepareResponseRecords()```, but for the moment I find this more like a hack than an actual implementation.
In the current approach, name compression is skipped for the first occurrence in answer section, and for the SRV as the spec requires, if SRV is to be appended. If it's fine not to compress the name for the moment, OK, if not I have to think what can be done further.

Reply:
>  If it's fine not to compress the name for the moment.

This sounds good to me.